File: line_2d3d/python/MyCamera.py
from typing import List
import numpy as np
from MyUtils import *

class MyCamera :

#   private PMatrix3D cam_mat;
  
  def __init__(self, cam_pos:List[float], look_at:List[float], up:List[float], zoom:float) :
    
    view_dir = np.subtract(cam_pos, look_at)
    self.view_dir = view_dir / np.linalg.norm(view_dir)

    side_dir = np.cross(up, view_dir)
    self.side_dir = side_dir / np.linalg.norm(side_dir)

    e = np.cross(view_dir, side_dir)
    self.e = e / np.linalg.norm(e)
    a1 = [self.side_dir[X], self.side_dir[Y], self.side_dir[Z], -np.dot(self.side_dir, cam_pos)]
    a2 = [self.e[X], self.e[Y], self.e[Z], -np.dot(self.e,cam_pos)]
    a3 = [self.view_dir[X], self.view_dir[Y], self.view_dir[Z], -np.dot(self.view_dir, cam_pos)]
    a4 = [0, 0, 0, 1]
    self.cam_mat = []
    self.cam_mat.append(a1)
    self.cam_mat.append(a2)
    self.cam_mat.append(a3)
    self.cam_mat.append(a4)
    # zoom = np.multiply(np.eye(4), zoom)
    self.cam_mat = np.multiply(self.cam_mat, zoom)
  
  def project(self, v3:List[float]) -> List[float]:
    v4 = list(v3)
    v4.append(0)
    v4_p = np.dot(self.cam_mat, v4)
    
    return v4_p[0:3]
  
  def console_print(self) :
   print("cam_mat ")                  
   print( f"  {self.cam_mat.m00:.1}  {self.cam_mat.m01:.1}  {self.cam_mat.m02:.1}  {self.cam_mat.m03:.1} "        )                
   print( f"  {self.cam_mat.m10:.1}  {self.cam_mat.m11:.1}  {self.cam_mat.m12:.1}  {self.cam_mat.m13:.1} "      )                  
   print( f"  {self.cam_mat.m20:.1}  {self.cam_mat.m21:.1}  {self.cam_mat.m22:.1}  {self.cam_mat.m23:.1} "          )               
   print( f"  {self.cam_mat.m30:.1}  {self.cam_mat.m31:.1}  {self.cam_mat.m32:.1}  {self.cam_mat.m33:.1} "   )
  
  


    

File: line_2d3d/cpp/line_2d3d/line_2d3d/MyCamera.cpp
#include "MyCamera.h"
using Eigen::Vector4d;


MyCamera::MyCamera(Vector3d cam_pos_arg, Vector3d look_at_arg, Vector3d up_arg, double zoom_arg)
{
	Vector3d view_dir = cam_pos_arg - look_at_arg;
	view_dir.normalize();
	Vector3d side_dir = up_arg.cross(view_dir);
	side_dir.normalize();
	Vector3d e = view_dir.cross(side_dir);
	e.normalize();

	cam_mat << side_dir[0], side_dir[1], side_dir[2], -side_dir.dot(cam_pos_arg),
		e[0], e[1], e[2], -e.dot(cam_pos_arg),
		view_dir[0], view_dir[1], view_dir[2], -view_dir.dot(cam_pos_arg),
		0, 0, 0, 1;

	cam_mat *= zoom_arg;
}

Vector3d MyCamera::project(Vector3d v3)
{
	Vector4d v4;
	v4.head(3) << v3;
	v4[3] = 0;
	Vector4d v4_p = cam_mat * v4;

	return v4.head(3);
}


//
//MatrixXd m(2, 2);
//m(0, 0) = 3;
//m(1, 0) = 2.5;
//m(0, 1) = -1;
//m(1, 1) = m(1, 0) + m(0, 1);
//std::cout << m << std::endl;
//Vector3d d3;
//d3 << 1, 2, 3;
//std::cout << d3 << std::endl;
//Vector4d d4;
//d4.head(3) << d3;
//d4[3] = 10;
//std::cout << d4 << std::endl;
